Improve reliability of findScrollableParent

GioSensation opened the pull request:
Improves reliability of findScrollableParent by accounting separately for horizontal and vertical overflows and also by checking for both `scroll` and `auto` values of the overflows, rather than just for `visible`. This fixes a bug that was unnecessarily scrolling items that had `overflow: hidden`. The cases included in index.html seem to be working just fine.

Note: I had closed a previous pull request with a similar fix, because that one was not really fixing the issue. This does, so here I go again.

GioSensation commented:
@kevincennis This pull request should fix the issue you were investigating about [here](https://github.com/iamdustan/smoothscroll/pull/75). Enjoy.

jeremenichelli commented on dist/smoothscroll.js:
```diff
@@ -99,22 +99,37 @@
+        hasVerticalScrollableSpace = el.clientHeight < el.scrollHeight;
+        hasHorizontalScrollableSpace = el.clientWidth < el.scrollWidth;
+        hasScrollableOverflowY =
+          w.getComputedStyle(el, null).overflowY === 'auto' ||
```
It would be good to create a test case for this, but I got the feeling that the `auto`check could case a false positive, not all `auto` overflow will have scrolling to do and this will prevent the script to scroll body.

jeremenichelli commented on dist/smoothscroll.js:
```diff
@@ -99,22 +99,37 @@
+          w.getComputedStyle(el, null).overflowY === 'scroll';
+        hasScrollableOverflowX =
+          w.getComputedStyle(el, null).overflowX === 'auto' ||
+          w.getComputedStyle(el, null).overflowX === 'scroll';
```
We are calling four times `getComputedStyle` here which could be memoized before all these individual checks.

jeremenichelli commented on src/smoothscroll.js:
```diff
@@ -93,22 +93,37 @@
+          w.getComputedStyle(el, null).overflowY === 'scroll';
+        hasScrollableOverflowX =
+          w.getComputedStyle(el, null).overflowX === 'auto' ||
+          w.getComputedStyle(el, null).overflowX === 'scroll';
```
Instead of repeating this logic I think is better to create a method which receives a node and returns true or false to avoid duplicated code.

jeremenichelli commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "title": "smoothscroll",
   "name": "smoothscroll-polyfill",
-  "version": "0.3.5",
+  "version": "0.3.6",
```
Avoid pushing versions. Just add changes to the src file, we will handle the releases on our end.

GioSensation commented:
@jeremenichelli Thanks for your review. I have updated the the repo with your suggestions and now it really looks much cleaner. I have added 3 methods in order to keep them easy to reason about and pretty straightforward. It adds a bit of clutter, but I believe it's better.

Please, check again.

Thanks.

jeremenichelli commented:
Awesome @GioSensation, can you just leave the file `src` in the diff? I gotta do a lot of gardening around so I can merge this and a couple more of pull request, plus feature addition for next release and it would simplify it on my end. 

If you don't have the time don't worry, I will probably cherry pick your additions and add them on my end. Thanks again for your work!

GioSensation commented:
@jeremenichelli I am not sure I get what you are asking. Do you want me to just leave the changes in `src/smoothscroll.js`, without adding `dist/smoothscroll.js` to the commit?

jeremenichelli commented:
Exactly that, primarily because I prefer to merge work instead of cherry picking when possible to visualize others collaboration, but if not it's OK.

GioSensation commented:
@jeremenichelli I have removed the changes in the dist file.

Hope it helps.

jeremenichelli commented:
Thanks @GioSensation. Sadly, this PR and others are entering in conflict (https://github.com/iamdustan/smoothscroll/pull/64/files) so I'm gonna manually pick up the work of each to simplify the release and be equally fair with everyone since someone would need to resolve the conflicts no matter the other I chose.

I'm gonna make a PR and put you as a reviewer so you can check your changes.

GioSensation commented:
No probs, @jeremenichelli. It's a bit of a bummer, but I am still happy that my contribution will make its way onto the release. Glad to help. I will wait for the notification from GitHub and then close this PR.

jeremenichelli commented:
@GioSensation wait, let me give it a try. I will merge the other PRs and see if I'm allowed to solve manually the conflicts. I would love if the actual contributions are part of the repository 😄

jeremenichelli commented:
Merged @GioSensation, sorry for the noise!
